sync/category.go

```diff
@@ -0,0 +1,93 @@
+package sync
+
+import (
+	"context"
+	"strconv"
+
+	"github.com/appuio/appuio-cloud-reporting/pkg/db"
+	"github.com/vshn/appuio-odoo-adapter/odoo/model"
+)
+
+// Reconcile synchronizes model.InvoiceCategory in Odoo based on the given db.Category according to the following rules:
+//  * If db.Category.Target is NULL then it will search for an existing category that matches the name:
+//    * If not found, it will create a new model.InvoiceCategory and set db.Category.Target to the ID returned by Odoo.
+//    * If found a match, it will use the model.InvoiceCategory and set db.Category.Target to the ID returned by Odoo.
+//  * If db.Category.Target has a value then it will search for a matching model.InvoiceCategory:
+//    * If not found, it will recreate the model.InvoiceCategory.
+//    * If found and model.InvoiceCategory.Name is up-to-date, it will return without error
+//    * If found and model.InvoiceCategory.Name differs from db.Category.Source, the model.InvoiceCategory is updated/reset.
+func (s *InvoiceCategoryReconciler) Reconcile(ctx context.Context, category *db.Category) error {
+	converted := model.InvoiceCategory{}
+	err := CategoryConverter{}.ToInvoiceCategory(category, &converted)
+	if err != nil {
+		return err
+	}
+	if !category.Target.Valid {
+		return s.findOrCreateCategory(ctx, category, converted)
+	}
+	return s.updateCategoryIfNeeded(ctx, category, converted)
+}
+
+func (s *InvoiceCategoryReconciler) findOrCreateCategory(ctx context.Context, category *db.Category, converted model.InvoiceCategory) error {
+	found, err := s.odoo.SearchInvoiceCategoriesByName(ctx, converted.Name)
+	if err != nil {
+		return err
+	}
+	if len(found) == 0 {
+		return s.createCategoryInOdoo(ctx, category)
+	}
+	category.Target.String, category.Target.Valid = strconv.Itoa(found[0].ID), true
+	return nil
+}
+
+func (s *InvoiceCategoryReconciler) createCategoryInOdoo(ctx context.Context, category *db.Category) error {
+	ic := newInvoiceCategory(nil)
+	err := CategoryConverter{}.ToInvoiceCategory(category, &ic)
+	if err != nil {
+		return err
+	}
+	created, err := s.odoo.CreateInvoiceCategory(ctx, ic)
+	if err != nil {
+		return err
+	}
+	category.Target.String, category.Target.Valid = strconv.Itoa(created.ID), true
+	return nil
+}
+
+func (s *InvoiceCategoryReconciler) updateCategoryIfNeeded(ctx context.Context, category *db.Category, ic model.InvoiceCategory) error {
+	// TODO: re Idempotency, should we search by name instead? Maybe the category in odoo got deleted & recreated with a new ID manually
```

Comment:
No I think we should definitely search by ID. If someone deletes and recreates a category, they created a new category and it is not the job of our operator to fix their mistake.